### PR TITLE
Update to 0.12.0

### PR DESCRIPTION
**Pull request by Southporter**
I took the liberty of updating the codebase to work with zig 0.12.0. There were a lot of changes due to std.os being largely moved to std.posix and some methods removed.

I've run the tests on each of the following:

| OS | x86_64 | Aarch64 |
| ---  | --- | --- |
| Linux | 100% | 100% |
| MacOS | N/A | 100% |
| Windows | 100% | 1 failed [^1] |

I know you have your #37 PR in progress. If you want to hold off on merging this, I would be happy to handle any merge conflicts that arise due to that PR.

[^1]: Windows Aarch64 has wasi `dangling_symlink` test failing on both 0.11.0 and 0.12.0.

### CODE REVIEW

**Comment by rdunnington** [comment]
Wow, this is a ton of work, thank you! I was planning on doing this after #37 as you mentioned. I would appreciate holding onto this until that one is merged since there are quite a few changes in there and I don't want to lose momentum on those changes. I'll ping this PR again when it's merged.

**Comment by rdunnington** [comment]
#37 is merged, so this is unblocked. Apologies for the conflicts. If you'd prefer, I can take the changes you've made so far and fix everything up, though it might take me a few days.

**Comment by Southporter** [comment]
I'll work on the conflicts. Most of them should be easy. I'll let you know if I have any questions.

**Comment by rdunnington** [comment]
Thanks for fixing all the conflicts. It looks like the last failing test is in the stable array lib. I have a repo for it [here ](https://github.com/rdunnington/zig-stable-array) and just finished upgrading it to 0.12.0, so you could try dropping in the latest version to see if it fixes anything. Otherwise I am happy to take it from here and investigate the last failing test, especially since it's in the mem64 stuff that just got added - it's possible there's a bug I missed that somehow the changes exposed.

File: src/vm_stack.zig
Comment (Southporter) [inline src/vm_stack.zig:2105]: For some reason, adding this debug line allows the mem64 tests to pass on Linux/windows/x86_64. Also running it with `-Doptimize=ReleaseSafe` without this statement allows it to pass as well. I think there might be a compilation issue, but need to dig into it a bit more.

File: src/vm_stack.zig
Comment (Southporter) [inline src/vm_stack.zig:2109]: This was the culprit. Old size was being pushed as an i32, but read in storeInMem as an i64. Put the offset way out of the range of the memory.

Comment (rdunnington) [reply src/vm_stack.zig:2109]: Ah good catch! I'll take a look at all the changes tonight and merge it soon. Thanks again for all the work.

File: build.zig.zon
Comment (rdunnington) [inline build.zig.zon:7]: I think this has changed since the `memory64` merge to be `test/wasm/main.zig`. Also we'd need to add:
* `test/wasi/main.zig`
* test/mem64/main.zig`

File: src/definition.zig
Comment (rdunnington) [inline src/definition.zig:3446]: Nice catch, thanks

File: src/vm_stack.zig
Comment (rdunnington) [inline src/vm_stack.zig:292]: I think this could be `@memcpy` since the stack spaces shouldn't overlap. Though it would be nice to have an assert double-checking that.

File: src/vm_stack.zig
Comment (rdunnington) [inline src/vm_stack.zig:346]: `@memcpy` + stack overlap assert?

File: src/vm_stack.zig
Comment (rdunnington) [inline src/vm_stack.zig:980]: I think this could just be `std.mem.readInt(read_type, mem, .little)`? I don't quite see the purpose of the `[0..byte_count]` part.

File: src/vm_stack.zig
Comment (rdunnington) [inline src/vm_stack.zig:1006]: Can this be `std.mem.readInt(read_type, mem[value_start..value_end], .little);`?

File: src/vm_stack.zig
Comment (rdunnington) [inline src/vm_stack.zig:1042]: Is `[0..byte_count]` needed?

File: src/vm_stack.zig
Comment (rdunnington) [inline src/vm_stack.zig:1080]: `@memcpy` + stack overlap assert?

File: src/vm_stack.zig
Comment (rdunnington) [inline src/vm_stack.zig:3442]: `@memcpy`?

File: src/vm_stack.zig
Comment (rdunnington) [inline src/vm_stack.zig:3482]: `@memcpy`?

File: src/vm_stack.zig
Comment (rdunnington) [inline src/vm_stack.zig:3544]: `@memcpy`?

File: src/vm_stack.zig
Comment (rdunnington) [reply src/vm_stack.zig:3482]: Actually scratch that, I think `memory.copy` can deal with overlapping memory regions, so `copyForwards` is the correct one here.

**Review by rdunnington — CHANGES_REQUESTED** [review]
I left comments on a few things that I'd like to see updated before the merge into `main`. I can also take over if you prefer, just let me know. Thanks!

File: src/vm_stack.zig
Comment (Southporter) [reply src/vm_stack.zig:980]: std.mem.readInt takes an array, so the slice needs to be comptime known. Currently `mem` here is a runtime slice, so the extra `[0..byte_count]` part is used to create the comptime known array length.

File: src/vm_stack.zig
Comment (Southporter) [reply src/vm_stack.zig:3442]: @memcpy works for only non-aliasing values. In some of these, they variables alias, so we have to use std.mem.copyForwards. I'll double check which ones alias and which ones don't and use `@memcpy` where possible.

File: src/vm_stack.zig
Comment (rdunnington) [reply src/vm_stack.zig:980]: Ah I see, that does make sense.

File: build.zig.zon
Comment (Southporter) [reply build.zig.zon]: Done.

File: src/vm_stack.zig
Comment (Southporter) [reply src/vm_stack.zig]: For this one, `@memcpy` isn't allowed. I've added an if check and copyBackwards because I was running into a failing wasm test where the dest_begin was greater than the source_begin.

File: src/vm_stack.zig
Comment (Southporter) [reply src/vm_stack.zig:352]: Added the assert, but `@memcpy` doesn't work here.

File: src/vm_stack.zig
Comment (Southporter) [reply src/vm_stack.zig:1087]: Added assert. `@memcpy` doesn't work here

**Review by rdunnington — APPROVED** [review]

**Pull request merged**
